[1.22] Volume populators redesign #1495 (#28610)
kubernetes/website: content/en/docs/concepts/storage/persistent-volumes.md

      storage: 10Gi
```

## Volume Populators and Data Sources

{{< feature-state for_k8s_version="v1.22" state="alpha" >}}

{{< note >}}

> Volume populars are available as alpha since Kubernetes 1.18, are redesigned in Kubernetes 1.22. The `AnyVolumeDataSource` feature must be enabled. Refer to the [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) documentation for more information.

Reviewer: ```suggestion
Volume populars are available as alpha since Kubernetes 1.18, and are redesigned in Kubernetes 1.22. The `AnyVolumeDataSource` feature must be enabled, to enable it please refer to the [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) documentation for more information.
```

Reviewer 2: BTW “populars” is a spelling error.

> Volume populars are available as alpha since Kubernetes 1.18, are redesigned in Kubernetes 1.22. The `AnyVolumeDataSource` feature must be enabled. Refer to the [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) documentation for more information.

Author: Thankss

{{< /note >}}

> Volume populators take advantage of a new PVC spec field called `dataSourceRef`. Unlike the

Author: Does this still apply if the intent is to deprecate the "old" field and require users to use the "new" field after this feature reaches GA status? We're adding dataSourceRef as a replacement for dataSource, but we can't tell people to stop using the old field and start using the new field until the normal alpha/beta/GA sequence.

> Volume populators take advantage of a new PVC spec field called `dataSourceRef`. Unlike the

Reviewer 2: It's OK to talk about deprecations once they're confirmed, but until then we avoid documenting the intent-to-deprecate.

It's also OK to express that, for clusters with the feature gate enabled, the `dataSourceRef` field is preferred. Notice how I wrote that? I could have written:
> It's also OK to express that, for clusters with the feature gate enabled, the new field is preferred.

(and actually I did, but I caught myself and reworded it).

> Volume populators take advantage of a new PVC spec field called `dataSourceRef`. Unlike the

Author: Understood, I'll fix this

> Volume populators take advantage of a new PVC spec field called `dataSourceRef`. Unlike the

Author: Reworded

`dataSource` field, which can only contain either a reference to another PersistentVolumeClaim
or to a VolumeSnapshot, the `dataSourceRef` field can contain a reference to any object in the

> same namespace, except for core objects other than PVCs.

Reviewer 3: Do core objects only refer to APIs defined in https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go?
Does this mean APIs defined in other groups under https://github.com/kubernetes/kubernetes/blob/master/pkg/apis such as StatefulSet in "apps" maybe supported?

Author: Just core. We won't reject one with a apiGroup of "apps" -- those will simply fail to bind because nobody will write a populator for statefulsets.

> ### Data Source Field Semantics

Reviewer 3: Currently under https://kubernetes.io/docs/concepts/storage/persistent-volumes/, we have "Volume Snapshot and Restore Volume from Snapshot Support" and "Volume Cloning" at the same level.  If we merge this new volumeDataSource with Volume Cloning, then we need to merge "Restore Volume from Snapshot" into the same heading as well.  Otherwise, it's better to have the 3 of them under different headings at the same level.

> ### Data Source Field Semantics

Reviewer 2: The concept we want to explain is that you can define a way to claim a PersistentVolume where that PV arrives already containing data. I'd make this be a second-level heading. Having introduced that generally, we should also enumerate how you can achieve that:
- cloning a PVC that already exists
- using a snapshot as the source for populating a PV using a `VolumeSnapshot` datasource (hyperlinking to https://kubernetes.io/docs/concepts/storage/volume-snapshots/ for more details).
- using a custom populator as the source for populating a PV

I'd make each of those be third-level headings under the general heading. This is because they describe different ways to end up with a ready-populated PV and, in fact, they use the same fields in the API.

Potentially, we could move all of these concepts to a new page about populating PVs during creation. For now I'd leave them in the Persistent Volumes topic.

> ### Data Source Field Semantics

Author: I like the idea of tabs and reworking this section but I am not good at layout design. My change was focused on adding the new information that's changed since the last release. I worry that by reworking everything we might lose existing information.

> ### Data Source Field Semantics

Reviewer 2: OK, that's fair enough. Let's focus on getting the content _accurate_ and then we can try to do a follow up PR that makes it more readable. Ideally we can get that follow-up PR in the dev-1.22 branch before the release, but it's not the end of the world if it goes in later.

@bswartz could you point out https://github.com/kubernetes/website/pull/28610#discussion_r677488539 in the PR description? This is to call out that we're taking a small shortcut with this PR, and to help frame how we review it.

> ### Data Source Field Semantics

Author: Updated

Use of the `dataSourceRef` field is nearly identical to the `dataSource` field, in that filling
in either one causes the other to be filled in with the same value, and neither can be changed
after creation. The two fields will thus always have the same contents. The two differences that

> matter are:

Reviewer: @sftim need an english native speaker to reword this, i feel like i need to think about this paragraph very hard to understand what it means but i'm also unable to come up with something better.

* The `dataSource` field ignores invalid values (as if the field was blank) while the
  `dataSourceRef` field never ignores values and will cause an error if an invalid value is
  used.

> * The `dataSourceRef` field many different types of objects, while the `dataSource` field only

Reviewer 3: Missing a verb after "The `dataSourceRef` field".

> * The `dataSourceRef` field many contain different types of objects, while the `dataSource` field

Reviewer 3: many contain -> may contain?

  only allows PVCs and VolumeSnapshots.

> Attempting to specify different values for the two fields is always an error.

Reviewer: ```suggestion
Attempting to specify different values for the two fields will result in an error.
```

Reviewer 2: The original wording feels closer to the style guide. I'm fine with either.

### Using Volume Populators

> Volume populators are controllers that can create non-empty volumes, where the contents of the

Author: Events, pods, configMaps -- anything with an empty apiGroup will fail at admission time. Things with non-empty apiGroup will be admitted but will sit there forever, unless there is a populator for that group/kind. In the latter case the data-source-validator will figure out if there's a populator or not and provide the feedback.

> Volume populators are controllers that can create non-empty volumes, where the contents of the

Reviewer 2: Great! Can we have that detail in the docs?

> Volume populators are controllers that can create non-empty volumes, where the contents of the

Author: Yes, I reworded this.

volume are determined by a Custom Resource. Users create a populated volume by referring to a
Custom Resource using the `dataSourceRef` field:

```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: populated-pvc
spec:
  dataSourceRef:
    name: example-name
    kind: ExampleDataSource

>     apiGroup: example.storage.k8s.io

Reviewer 3: May be better to use an external apiGroup as an example because core API groups except for PVC are not supported?

Author: The only one that's not supported is the "" (empty) apiGroup. I'm happy to change it to something without "k8s" in it though.

Reviewer 3: It's fine to leave it as it.

  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 10Gi
```

### Volume Populators Errors

> Because Volume Populators are external components, attempts to create a PVC that uses one

Reviewer 3: attempts -> attempt

Author: I think the plural version is actually correct here

Reviewer 3: Ok

> can fail if not all the correct components are installed. There is a data-source-validator

Reviewer 3: Add a link to the data-source-validator repo?

Author: This is user docs, not developer docs. I'm not sure it' makes sense to point to source code from here.

Reviewer 3: That's fine.

controller that will generate warning events on the PVC in the case that no populator
is registered to handle that kind of data source. If a populator is installed, any errors

> related to creation of the volume will reported by the populator controller itself as

Reviewer 3: will be reported

> events on the PVC.

Author: It's not built-in. I understand the desire to not document out of tree things here, but for Snapshots -- that's all out of tree. This isn't much different. It will be hard to describe what the plan is without at least gesturing towards the out of tree components.

> events on the PVC.

Reviewer 2: Out of tree stuff is always fine to document so long as it's Kubernetes code. If it's somewhere inside https://github.com/kubernetes or https://github.com/kubernetes-sigs then we can document it with zero qualms.

> events on the PVC.

Author: https://github.com/kubernetes-csi too?

> events on the PVC.

Reviewer 2: Ah yep, that too.

> events on the PVC.

Reviewer 2: We could write
```markdown
You can install the alpha [volume data source validator](https://github.com/kubernetes-csi/volume-data-source-validator)
controller into your cluster. That controller generates warning Events on a PVC in the case that no populator
is registered to handle that kind of data source. When a suitable populator is installed for a PVC, it's the
responsibility of that populator controller to report Events that relate to volume creation and issues during
the process.
```
How does that sound?

      storage: 10Gi
```

## Volume populators and data sources

{{< feature-state for_k8s_version="v1.22" state="alpha" >}}

{{< note >}}
Kubernetes supports custom volume populators; this alpha feature was introduced
in Kubernetes 1.18. Kubernetes 1.22 reimplemented the mechanism with a redesigned API.
Check that you are reading the version of the Kubernetes documentation that matches your
cluster. {{% version-check %}}
To use custom volume populators, you must enable the `AnyVolumeDataSource`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) for
the kube-apiserver and kube-controller-manager.
{{< /note >}}

Volume populators take advantage of a PVC spec field called `dataSourceRef`. Unlike the
`dataSource` field, which can only contain either a reference to another PersistentVolumeClaim
or to a VolumeSnapshot, the `dataSourceRef` field can contain a reference to any object in the
same namespace, except for core objects other than PVCs. For clusters that have the feature

> gate enabled, use of the `dataSourceRef` is preferred over `dataSource`.

Reviewer 2: Yep, happy with this wording! Thanks.

## Data source references

The `dataSourceRef` field behaves almost the same as the `dataSource` field. If either one is
specified while the other is not, the API server will give both fields the same value. Neither
field can be changed after creation, and attempting to specify different values for the two
fields will result in a validation error. Therefore the two fields will always have the same
contents.

There are two differences between the `dataSourceRef` field and the `dataSource` field that
users should be aware of:

> * The `dataSource` field ignores invalid values (as if the field was blank) while the
>   `dataSourceRef` field never ignores values and will cause an error if an invalid value is
>   used. Invalid values are any core object (objects with no apiGroup) except for PVCs.

Reviewer 2: Is this correct:
```suggestion
   For the `dataSource` field, you can only set the `apiGroup` field to `""` (the empty string,
   representing the core {{< glossary_tooltip text="API Group" term_id="api-group" >}})
   if you specify its `kind` as `PersistentVolumeClaim`; if you specify a different `kind` and
   set `apiGroup` to `""` then the API server rejects the request.
   For the `dataSourceRef` field you can specify any `kind` and `apiGroup`; however, the
   only kind within the core API group that can act as a data source is still PersistentVolumeClaim.
   If you try to create a PersistentVolumeClaim using some other core API as a data source,
   the PersistentVolumeClaim you create is invalid.
```
? I couldn't be sure I understood the text I've tried to reword.

* The `dataSourceRef` field may contain different types of objects, while the `dataSource` field

>   only allows PVCs and VolumeSnapshots.

Reviewer 2: What happens if you try to specify a different `dataSource` kind (specifying an API group other than `""`)?

### Using volume populators

Volume populators are {{< glossary_tooltip text="controllers" term_id="controller" >}} that can
create non-empty volumes, where the contents of the volume are determined by a Custom Resource.

> Users create a populated volume by referring to a Custom Resource using the `dataSourceRef` field:

Reviewer 2: ```suggestion
You create a populated volume by setting the `dataSourceRef` field to refer to a
{{< glossary_tooltip term_id="CustomResourceDefinition" text="custom resource" >}}:
```
